### ISSUE TITLE
aruco python binding: refineDetectedMarkers

### ISSUE THREAD

**Issue by ShT3ch** [Itseez/opencv_contrib#692]
opencv-modules commithash: ba1d3ef99cf5e67241c5a31dbd9c344d12a2f7c5
opencv commithash: f6fa1cee2b9fc726452779cfd484ac382c142bab
Windows 8 x64
vc14 x64 toolchain Release mode
python 27 x64
numpy==1.9.3 x64

Code below causes exception 
```
more_corners, more_ids, rej, recovered_ids = cv2.aruco.refineDetectedMarkers(img, board, corners, ids, rejected)
TypeError: rejectedCorners is not a numpy array, neither a scalar

```
```
import cv2
import numpy as np

tag_size_m = 0.058
tag_margin_m = 0.012
board = cv2.aruco.GridBoard_create(2, 4, tag_size_m, tag_margin_m, cv2.aruco.getPredefinedDictionary(4))

cap = cv2.VideoCapture(0)

while True:
    _, img = cap.read()

    corners, ids, rejected = cv2.aruco.detectMarkers(img, board.dictionary)

    if 0 < len(corners) < len(board.objPoints):
        more_corners, more_ids, rej, recovered_ids = cv2.aruco.refineDetectedMarkers(img, board, corners, ids, rejected)

    k = cv2.waitKey(1) & 0xff
    if k == 27:
        break

```

If I reassign `rejected ` as `rejected = np.array(rejected)` it throws exception:

```
Traceback (most recent call last):
    more_corners, more_ids, rej, recovered_ids = cv2.aruco.refineDetectedMarkers(img, board, corners, ids, rejected)
cv2.error: ......\opencv\modules\core\src\matrix.cpp:553: error: (-215) r == Range::all() || (0 <= r.start && r.start < r.end && r.end <= m.size[i]) in function cv::Mat::Mat
```
shape of rejected in that case is (13L, 1L, 4L, 2L)

Any of my attempts to fix it by arrays ended with same exception. I've also tried: 
`rejected = np.reshape(np.array(rejected), (len(rejected) * 4, 2))`
`rejected = np.array(rejected).squeeze()`

Structure of `rejected` is similar with structure of `corners`.
But in c++ they have different type in signature of refineDetectedMarkers, is it ok?
```
void refineDetectedMarkers(InputArray _image, Ptr<Board> &_board,
                           InputOutputArrayOfArrays _detectedCorners, InputOutputArray _detectedIds,
                           InputOutputArray _rejectedCorners, InputArray _cameraMatrix,
                           InputArray _distCoeffs, float minRepDistance, float errorCorrectionRate,
                           bool checkAllOrders, OutputArray _recoveredIdxs,
                           const Ptr<DetectorParameters> &_params)
```

How I can handle that?

**Comment by mshabunin** [comment]
`detectMarkers` returns `rejected` as a list of np.array, corresponding to OutputArrayOfArrays in C++. `refineDetectedMarkers` accepts `rejected` as an InputOutputArray in C++ or single np.array in Python.

So, I guess, you should pass each element of the rejected list to the refine method.

**Comment by ShT3ch** [comment]
I've tried that:
```
    if 0 < len(corners) < len(board.objPoints):
        for rejected in rejected:
            more_corners, more_ids, rej, recovered_ids = cv2.aruco.refineDetectedMarkers(img, board, corners, ids, rejected)
```
There is exaclty the same exception: 
`cv2.error: ......\opencv\modules\core\src\matrix.cpp:553: error: (-215) r == Range::all() || (0 <= r.start && r.start < r.end && r.end <= m.size[i]) in function cv::Mat::Mat`

Line changed because openCV was pulled and rebuilded.

Moreover, in refineDetectedMarkers c++ code there exist line:
`Point2f rejCorner = _rejectedCorners.getMat(j).ptr< Point2f >()[(c + k) % 4];`
that tells that expected shape of rejectedCorners is (n,4,2), isn't it?
https://github.com/Itseez/opencv_contrib/blob/master/modules/aruco/src/aruco.cpp#L1168

**Comment by mshabunin** [comment]
Strange, I tried to run same script without any problem (though I used static board sample image from aruco tutorial instead of camera feed):
```.py
for one in rejected:
    more_corners, more_ids, rej, recovered_ids = cv2.aruco.refineDetectedMarkers(img, board, corners, ids, one)
```
Try to check `type(rejected)` before the function call. If it is np.array - print `np.shape(rejected)`.

**Comment by mshabunin** [comment]
Or maybe `_rejectedCorners` C++ type should be changed to `InputOutputArrayOfArrays`? Can you check it?

**Comment by ShT3ch** [comment]
```
        print type(rejected)
        print len(rejected)
        print type(rejected[0])
        print rejected[0].shape
```
```
<type 'list'>
9
<type 'numpy.ndarray'>
(1L, 4L, 2L)
```
Its totally the same structure as `corners`

> Or maybe _rejectedCorners C++ type should be changed to InputOutputArrayOfArrays? Can you check it?

I've tried that. Errors `TypeError: rejectedCorners is not a numpy array, neither a scalar` stopped when passing "pure" `rejected` , but wrong indexation error remained in place.

**Comment by mshabunin** [comment]
@ShT3ch , please check the fix in #696.

**Comment by ShT3ch** [comment]
It works!
And now I understand why my fix with exchange `InputOutputArray `-> `InputOutputArrayOfArrays ` dont work: I dont change signature in .hpp
